src/intervals/construction.jl

```diff
@@ -0,0 +1,216 @@
+Interval(a::T, b::S) where {T<:AbstractFloat, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::T) where {T<:AbstractFloat} = Interval{T}(a, b)
+
+Interval(a::T, b::S) where {T, S} = Interval{promote_type(default_bound(), T, S)}(a, b)
+Interval(a::T, b::T) where T = Interval{promote_type(default_bound(), T)}(a, b)
+
+#= Irrational =#
+# Single argument Irrational constructor are in IntervalArithmetic.jl
+# as generated functions need to be define last.
+Interval(a::Irrational, b::T) where {T<:AbstractFloat} = Interval{T}(T(a, RoundDown), b)
+Interval(a::T, b::Irrational) where {T<:AbstractFloat} = Interval{T}(a, T(b, RoundUp))
+
+function Interval(a::Irrational, b::Irrational)
+    T = default_bound()
+    return Interval{T}(a, b)
+end
+
+#= Interval =#
+Interval{T}(x::Interval{T}) where T = x
+Interval{T}(x::Interval) where T = Interval{T}(x.lo, x.hi)
+
+#= Complex =#
+Interval(x::Complex) = Interval(real(x)) + im*Interval(imag(x))
+
+# These definitions has been put there because generated functions must be
+# defined after all methods they use.
+@generated function Interval{T}(x::Irrational) where T
+    res = Interval{T}(x(), x())  # Precompute the interval
+    return :(return $res)  # Set body of the function to return the precomputed result
+end
+
+Interval(x::Irrational) = Interval{default_bound()}(x)
+
+"""
+    interval(a, b)
+
+Create an interval, checking whether [a, b] is a valid `Interval`
+If so, then an `Interval(a, b)` object is returned;
+if not, a warning is printed and the empty interval is returned.
+"""
+function interval(a::T, b::S) where {T<:Real, S<:Real}
+    if !is_valid_interval(a, b)
+        @warn "Invalid input, empty interval is returned"
+        return emptyinterval(promote_type(T, S))
+    end
+
+    return Interval(a, b)
+end
+
+interval(a::Real) = interval(a, a)
+
+const checked_interval = interval
+
+"""
+    a..b
+    ..(a, b)
+
+Create the interval `[a, b]`.
+
+Validity of the interval is checked, but nothing is done to compensate for
+the fact that floating point literals are rounded to nearest when parsed.
+
+Use the string macro `I"[a, b]"` to ensure tight enclosure around the number
+that is typed in, even when it is not exactly representable as a floating point
+number (like `0.1`).
+"""
+..(a, b) = checked_interval(a, b)
+
+"""
+    a ± b
+
+Create the interval `[a - b, a + b]`.
+
+Despite using the center-radius notation for its creation, the interval is
+still represented by its bounds internally.
+"""
+a ± b = checked_interval(-(a, b, RoundDown), +(a, b, RoundUp))
+±(a::Interval, b) = Interval(-(a.lo, b, RoundDown), +(a.hi, b, RoundUp))
+
+"""
+    atomic(::Type{<:Interval}, x)
+
+Construct the tightest interval of a given type that contains the value `x`.
+
+If `x` is an `AbstractString`, the interval is created by calling `parse`.
+
+If `x` is an `AbstractFloat`, the interval is widen to two eps to be sure
+to contain the number that was typed in.
```

Comment:
If `x::AbstractFloat`, the interval produced is not the tightest interval containing the value `x` (first line of the docstrings). The following illustrates my point:
```julia
julia> dump( IntervalArithmetic.atomic(Interval{Float64}, 0.1) )
Interval{Float64}
  lo: Float64 0.09999999999999999
  hi: Float64 0.10000000000000002

julia> dump( I"[0.1]" )
Interval{Float64}
  lo: Float64 0.09999999999999999
  hi: Float64 0.1
```
Clearly, the second interval is tighter than the first, and contains (the mathematical real) 0.1.

Reply:
Let me clarify my point: Both intervals `I"0.1"` and `IntervalArithmetic.atomic(Interval{Float64}, 0.1)` contain/include the mathematical 1//10, as well as the Float64 `0.1` (which as we know is slightly larger than 1//10, but the first one is tighter.

Reply:
The awkward thing here is that when `atomic` recieves `0.1`, we have no way to know if the user meant `1//10`, or actually meant `0.1`, or `10000000000000001//10000000000000000`. `atomic(Interval{Float64}, 0.1)` is the tightest interval that contain both `1//10` and `10000000000000001//10000000000000000`, both being something the user could have meant before we get `0.1` in the function.

What the previous implementation was doing was return the tighest interval that contain `rationalize(x)`. I think it is fine if we document it properly, but it is not what `atomic` was documented to do.

I hoped that with a macro, we could get the string representation of decimal number literal, but it is apparently not the case, see #502 .

Reply:
I think a good starting point is to assume that `0.1` is a floating point number (because that's what the computer interprets), and not necessarily the mathematical value corresponding to `1//10`. 

An analogy could be `BigFloat`s behavior: `BigFloat(0.1)` refers to the bigfloat corresponding to the floating point representation of `0.1`, while `BigFloat(1//10)` or `BigFloat("0.1")` refer to the nearest bigfloat point corresponding to `1//10`, for the specific precision used in the computation.